### PR TITLE
Move delete by query helpers into core

### LINKED ISSUE #22616: Move DeleteByQuery helper classes from reindex module to core
today if a plugin needs DBQ functionality they include / depend on the reindex jar. This is really bad since it's not intended to be a dependency of another plugin. Yet, DBQ is useful in some situations and it can easily be used without all the request / response classes. We should have a static helper class that does all the heavy lifting for us in core and them do the exposure in the reindex module.

@nik9000 can you take a stab at this?

### PR DESCRIPTION
**Pull request by nik9000**
This moves the building blocks for delete by query into core. This
should enabled two things:
1. Plugins other than reindex to implement "bulk by scroll" style
operations.
2. Plugins to directly call delete by query. Those plugins should
be careful to make sure that task cancellation still works, but
this should be possible.

Notes:
1. I've mostly just moved classes and moved around tests methods.
2. I haven't been super careful about cohesion between these core
classes and reindex. They are quite interconnected because I wanted
to make the change as mechanical as possible.

Closes #22616

### CODE REVIEW

**Comment by nik9000** [comment]
@tlrx, can you have a look at this? I'd love any suggestions on how to make this less messy!

File: core/src/main/java/org/elasticsearch/action/bulk/byscroll/package-info.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * Infrastructure for actions that modify documents based on the results of a scrolling query.
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/action/bulk/byscroll/package-info.java]: +1

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/RoundTripTests.java [hunk 2]
```diff
@@ -199,39 +190,6 @@ private void assertRequestEquals(AbstractBulkByScrollRequest<?> request, Abstrac
...
-        assertResponseEquals(response, tripped);
-    }
-
-    public void testBulkIndexByScrollResponse() throws IOException {
-        BulkIndexByScrollResponse response = new BulkIndexByScrollResponse(timeValueMillis(randomNonNegativeLong()), randomStatus(),
-                randomIndexingFailures(), randomSearchFailures(), randomBoolean());
-        BulkIndexByScrollResponse tripped = new BulkIndexByScrollResponse();
-        roundTrip(response, tripped);
-        assertResponseEquals(response, tripped);
-    }
```
Comment (tlrx) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/RoundTripTests.java]: +1 too - I never noticed these tests...

**Review by tlrx — APPROVED** [review]
LGTM, thanks @nik9000

**Comment by tlrx** [comment]
> I'd love any suggestions on how to make this less messy!

Suggestions are easy but I'm not sure how feasible they are :) I'd start by trying to extract `RequestWrapper`, `ScriptApplier` along with `copyMetadata() copyRouting()` methods from `AbstractAsyncBulkByScrollAction`. I'd love to have a kind of ScrollingProcessor (and SlicedScrollingProcessor) piloted by the AsyncBulkByScrollAction and combined with a BulkProcessor...

**Comment by nik9000** [comment]
Thanks for reviewing @tlrx! I'm not really sure what you mean about moving things around. I'll merge and you can open a PR to show me if you like?

master: 2e48fb8294a924bc3742419fe80e4a29fdfa0f54
5.x:

**Comment by s1monw** [comment]
@polyfractal just FYI this should help you using DBQ from a plugin /cc @jasontedor

**Comment by polyfractal** [comment]
<3 Thanks @nik9000!

**Comment by javanna** [comment]
Seems like the docs should have been updated after this change, see https://github.com/elastic/elasticsearch/issues/23845#issuecomment-291201807

**Pull request merged**